getwax/wax
# Process emails and initiate recovery

This PR is dependent on https://github.com/getwax/wax/pull/199, merge that first

## What is this PR doing?

- Create sqlite db and save emails
- Mocks processing of emails
- Initiates recovery on recovery plugin 
- Adds script to set up recovery for relayer - `setupRecoveryForRelayer.ts`
- Adds basic formatting and linting

## Happy path flow
1. Once an email is fetched via `EmailService.pollEmails()`, the app saves it to a the db and emits and event.
2. This event will trigger `EmailService.processEmails()` which finds eligible emails from the db, generates the circuit inputs (mocked for now), and calls the ethereum service.
3. `EthereumService.initiateRecovery()` is called which executes the `initiateRecovery` function on the recovery plugin
4. The email service will then mark the email as processed in the db

## How can these changes be manually tested?
Testing these changes is quite manual at the moment, can definitely streamline this in future work. Instead of running this manually yourself, you can watch a demo [here](https://www.loom.com/share/b0829cc3b933420495756b25937ea307?sid=0ad92f1d-4467-49da-966d-f2f5a0d52650).

1. Run `yarn setup` in `/zkp/relayer` and add your email details to the config file - I created a new gmail account specifically for testing. Creating an app password was the quickest option to get things working. You can google how to do that but you need to set up mfa before you can generate one. You the generated app password as the password for your gmail account in the config file.
2. Start a local hardhat node with `./script/start-hardhat.sh` in `account-integrations/safe`
3. Start a local bundler. `yarn run bundler --unsafe` if using eth infinitisms (need to do this automatically in previous script)
4. Set up recovery by running `yarn hardhat run ./script/setupRecoveryForRelayer.ts` in `account-integrations/safe`.
5. Send an email to your chosen email address (subject etc doesn't matter for now - parsing the email fully will be an upcoming PR)
6. Copy `SafeZkEmailRecoveryPlugin.json` from `account-integrations/safe/out` (might need to build contracts with `forge build`), paste it in `account-integrations/zkp/relayer/src/services/`. This is for the abi.
7. Start the node app with `yarn dev` in `zkp/relayer`. The email should be picked up and recovery should be initiated automatically. You should see this log printed `Recovery initiated by: ${emails[i].sender}`

## Does this PR resolve or contribute to any issues?
Contributes to #173

## Checklist

## Guidelines
- If your PR is not ready, mark it as a draft
- The `resolve conversation` button is for reviewers, not authors
  - (But add a 'done' comment or similar)
